README


This archive contains the distribution Env-C,
version 0.13:

  Get/Set/Unset Environment Variables on the C level

This software is copyright (c) 2002 by Michael Schout.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.015.



Changes
Revision history for Perl extension Env::C.

{{$NEXT}}
    - fix three different test failures on Windows (github #2)
    - Add CAVEATS section to POD and explain how getenv() and %ENV differ on
      Windows.
    - skip t/leak.t on cygwin – setenv() in cygwin is known to leak memory
    - Win32: use _putenv_s() to implement a setenv() wrapper
    - Win32: use _putenv_s() to implement unsetenv()
    - fix memory leak typo for systems without setenv()

0.15  2017-08-11
    - make signture test an AUTHOR test
    - Fix some compiler warnings [Thanks Petr Písař]
    - Add setenv_multi() and unsetenv_multi() for bulk Env operations [Thanks
      Sergey Panteleev]

0.14  2017-03-31
    - minor POD updates [Thanks José Joaquín Atria]
    - Add tests for override flag [Thanks José Joaquín Atria]
    - Use is() instead of ok() in tests [Thanks José Joaquín Atria]
    - apply skiplist to signature test

0.13  2015-05-02
    - flip order of ppport.h and XSUB.h includes to fix duplicate declaration
      warnings [Thanks Ian Bottomley]
    - Fix test failures in docker containers because smoke.t assumes $ENV{USER}
      exists, but in docker this is not (necessarily) true.

0.12  2014-11-03
    - Narrow down "panic: free from wrong pool" error to perl with both threads,
      and PERL_TRACK_MEMPOOL (enabled implicitly with DEBUGGING).  work around
      by using PL_use_safe_putenv only if building under threaded perl with
      PERL_TRACK_MEMPOOL.  [#99962]

0.11  2014-10-31
    - FreeBSD: fix SIGV at exit for perl < 5.19.6 by using PL_use_safe_putenv.
      This causes setenv to leak memory, but seems to be the only way to avoid
      SIGV. [#49872]
    - MANIFEST.SKIP: add MYMETA.json, MYMETA.yml
    - update to latest ppport.h
    - prereqs: require Test::More v0.88 or later, exclude optional
      Module::Signature

0.10  2013-04-09
    - fix memory leak due to PL_use_safe_putenv = 1 [#49872]

0.09  2012-06-25
    - new maintainer: Michael Schout <mschout@cpan.org>
    - use Dist::Zilla for distribution management
    - fix: getallenv() segmentation fault under OS X 10.7

0.08 Thu Sep  4 08:50:40 PDT 2008
    - fix the first test, where getenv returns undef on windows, rather than
      ''.
    - undo part of the patch for windows in wez@messagesystems.com's patch
    - another patch from wez@messagesystems.com, to solve the "panic: free from
      wrong pool during global destruction" problem he detected on his x64
      linux build. using PL_use_safe_putenv.

0.07 Sun Aug 31 15:14:09 PDT 2008
    - Env-C assumes that there are no native functions for manipulating the
      environment on Solaris 10, and this results in a segmentation fault when
      running the test script.

      The attached patch detects solaris versions that have these functions in
      libc by doing the same thing that the stdlib.h header does to reveal the
      prototype.  Ideally the Makefile.PL would detect the functions using a
      compiler check, but this seems like the simplest approach for the moment.
      [Patch by wez@messagesystems.com]
    - include the latest ppport.h

0.06 Mon May 23 18:53:52 EDT 2005
    - As AIX has no unsetenv, reuse the code for Solaris to implement it [Larry
      Pells <lpells@spillman.com>]

0.05 Thu May 19 15:05:37 EDT 2005
    - add Solaris support [Larry Pells <lpells@spillman.com>]

0.04 Wed Jun 23 17:52:38 IDT 2004
    - add the missing license info
    - add MSWin32 support [Larry Pells <lpells@spillman.com>]

0.03 Sun Mar  9 10:20:56 EST 2003
    - refuse to build for solaris and MSWin32, since it won't work
    - add the missing include #include <stdlib.h>

0.02 Sat Mar  8 22:20:56 EST 2003
    - fix doc typo [Mark James <mrj@bigpond.net.au>]
    - add a warning that the app is not thread-safe

0.01  Sat Aug 17 18:31:45 2002
    - first version



test.pl
use strict;
#use warnings;

use Test::More tests => 5;
use Env::C;

# getenv
my $key = "USER";
my $val_orig = Env::C::getenv($key);
is $val_orig, $ENV{$key}, "getenv matches perl ENV for $key";

# unsetenv
diag "unsetting an env";
Env::C::unsetenv($key);
diag "getting it";
my $val = Env::C::getenv($key);
is $val, undef, "$key is no longer set in C env";

# setenv
my $val_new = "foobar";
Env::C::setenv($key, $val_new);
diag "called setenv";
$val = Env::C::getenv($key) || '';
print "# [$key] expecting '$val_new', got '$val'\n";
ok $val eq $val_new ? 1 : 0;

# restore
Env::C::setenv($key, $val_orig);
$val = Env::C::getenv($key) || '';
print "# [$key] expecting '$val_orig', got '$val'\n";
ok $val eq $val_orig ? 1 : 0;

my $env = Env::C::getallenv();
print "# ", scalar(@$env), " env entries\n";
#print join "\n", @$env;
ok @$env;


C.pm
package Env::C;

require 5.005;

use strict;
#use warnings;

require DynaLoader;
@Env::C::ISA = qw(DynaLoader);
$Env::C::VERSION = '0.08';

bootstrap Env::C $Env::C::VERSION;

1;
__END__

=head1 NAME

Env::C - Get/Set/Unset Environment Variables on the C level



=head1 SYNOPSIS

  use Env::C;
  
  my $key = "USER";
  $val = Env::C::getenv($key) || '';
  
  Env::C::setenv($key, "foobar", [$override]);
  $new_val = Env::C::getenv($key) || '';
  
  Env::C::unsetenv($key);
  
  my $ar_env = Env::C::getallenv();
  print join "\n", @$ar_env;



=head1 DESCRIPTION

This module provides a Perl API for getenv(3), setenv(3) and
unsetenv(3). It also can return all the C<environ> variables.

Sometimes Perl invokes modules with underlaying C APIs which rely on
certain environment variables to be set, if these variables are set in
Perl and the glue code doesn't worry to set them on the C level, these
variables might not be seen by the C level. This module shows what
really the C level sees.




=head2 FUNCTIONS

=over

=item * getenv()

  $val = Env::C::getenv($key);

Returns the value of the environment variable matching the key or
C<undef>.

=item * setenv()

  Env::C::setenv($key, $value, [$override]);

The setenv() function adds the variable C<$key> to the environment
with the value C<$value>, if C<$key> does not already exist.  If
C<$key> does exist in the environment, then its value is changed to
C<$value> if C<$override> is non-zero; if C<$override> is zero or is
not passed, then the value of C<$key> is not changed.

=item * unsetenv()

  Env::C::unsetenv($key);

The unsetenv() function deletes the variable C<$key> from the
environment.

=item * getallenv()

  my $ar_env = Env::C::getallenv();
  print join "\n", @$ar_env;

The getallenv() function returns an array reference which includes all
the environment variables.

=back




=head2 EXPORT

None.




=head1 Thread-safety and Thread-locality

This module should not be used in the threaded enviroment.

Thread-locality: the OS, which maintains the struct C<environ>, shares
it between all threads in the process. So if you modify it in one
thread, all other threads will see the new value. Something that will
most likely break the code.

This module is not thread-safe, since two threads may attempt to
modify/read the struct C<environ> at the same time. I could add
locking if in threaded-environment. However since the lock can't be
seen by other applications, they can still bypass it causing race
condition. But since thread-locality is not maintained, making this
module thread-safe is useless.

If you need to modify the C level of C<%ENV> for all threads to see,
do that before threads are started. (e.g. for mod_perl 2.0, at the
server startup).




=head1 AUTHOR

Stas Bekman E<lt>stas@stason.orgE<gt>



=head1 COPYRIGHT

This is a free software; you can redistribute it and/or modify it
under the terms of the Artistic License.



=head1 SEE ALSO

L<perl>.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.015.
use strict;
use warnings;

use 5.005;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Get/Set/Unset Environment Variables on the C level",
  "AUTHOR" => "Michael Schout <mschout\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Env-C",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Env::C",
  "PREREQ_PM" => {
    "DynaLoader" => 0,
    "Test::More" => "0.88",
    "strict" => 0
  },
  "TEST_REQUIRES" => {
    "Socket" => 0,
    "Test::More" => "0.88"
  },
  "VERSION" => "0.13",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "DynaLoader" => 0,
  "Socket" => 0,
  "Test::More" => "0.88",
  "strict" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



